File: lib/iriq/reducer.rb
module Iriq
  # Reducers consume the Event stream emitted by Corpus#observe and update
  # the storage backend's materialized views. Each Reducer is a callable
  # that takes (event, storage) and applies the appropriate storage
  # operation; non-applicable events are no-ops via the EVENT_TYPES gate.
  #
  # Adding a new metric is: define a new Event subtype, write a Reducer that
  # handles it, register it in DEFAULTS — no other module changes.
  module Reducer
    # Each entry: { event_class => [lambda(event, storage) -> result] }.
    # Lambdas may return the result of the underlying storage call so
    # callers (Corpus#observe) can pick up the cluster they need to return.
    DEFAULTS = {
      Event::HostSeen        => [->(e, s) { s.increment_host(e.host) }],
      Event::PathLengthSeen  => [->(e, s) { s.increment_path_length(e.length) }],
      Event::RawShapeSeen    => [->(e, s) { s.increment_raw_shape(e.shape) }],
      Event::FingerprintSeen => [->(e, s) { s.increment_fingerprint(e.shape) }],
      Event::PositionSeen    => [->(e, s) { s.observe_position(e.position, e.value, e.type) }],
      Event::ClusterAddition => [->(e, s) { s.add_to_cluster(e.key, e.host, e.scheme, e.shape, e.identifier) }],
    }.freeze

    module_function

    # Apply the event to the storage via all Reducers registered for its
    # type. Returns the last non-nil reducer result — used by Corpus#observe
    # to pick up the Cluster created/updated by Event::ClusterAddition.
    def apply(event, storage, reducers: DEFAULTS)
      results = reducers.fetch(event.class, [])
      result  = nil
      results.each do |r|
        rv = r.call(event, storage)
        result = rv unless rv.nil?
      end
      result
    end
  end
end

File: go/reducer.go
package iriq

// Reducer consumes a single Event and applies the corresponding mutation
// to the storage backend. Reducers are independent: adding a new
// materialized view is "define a new Event type, write a Reducer that
// handles it, register it via DefaultReducers" — no other module changes.
//
// The result is the (optional) value the reducer wants to surface to the
// caller — Corpus.Observe uses it to pick up the *Cluster created by
// EventClusterAddition.
type Reducer func(e Event, s Storage) any

// DefaultReducers maps each Event type's EventKind() to the registered
// reducer chain. Build via init() so the table is constructed once.
var DefaultReducers = map[string][]Reducer{
	(EventHostSeen{}).EventKind(): {func(e Event, s Storage) any {
		s.IncrementHost(e.(EventHostSeen).Host)
		return nil
	}},
	(EventPathLengthSeen{}).EventKind(): {func(e Event, s Storage) any {
		s.IncrementPathLength(e.(EventPathLengthSeen).Length)
		return nil
	}},
	(EventRawShapeSeen{}).EventKind(): {func(e Event, s Storage) any {
		s.IncrementRawShape(e.(EventRawShapeSeen).Shape)
		return nil
	}},
	(EventFingerprintSeen{}).EventKind(): {func(e Event, s Storage) any {
		s.IncrementFingerprint(e.(EventFingerprintSeen).Shape)
		return nil
	}},
	(EventPositionSeen{}).EventKind(): {func(e Event, s Storage) any {
		ev := e.(EventPositionSeen)
		s.ObservePosition(ev.Position, ev.Value, ev.Type)
		return nil
	}},
	(EventClusterAddition{}).EventKind(): {func(e Event, s Storage) any {
		ev := e.(EventClusterAddition)
		return s.AddToCluster(ev.Key, ev.Host, ev.Scheme, ev.Shape, ev.Identifier)
	}},
}

// ApplyEvent runs the registered reducers for the event's kind against
// the storage. Returns the last non-nil reducer result (used by
// Corpus.Observe to pick up the *Cluster from EventClusterAddition).
func ApplyEvent(e Event, s Storage) any {
	var result any
	for _, r := range DefaultReducers[e.EventKind()] {
		if rv := r(e, s); rv != nil {
			result = rv
		}
	}
	return result
}
